### PR TITLE
Remove race condition when accessing remote bulker map

### LINKED ISSUE #4170: Test_Agent_Remote_ES_Output flaky due to race
## Flaky Test

* **Test Name:** Test_Agent_Remote_ES_Output
* **Link:** https://github.com/elastic/fleet-server/blob/main/internal/pkg/bulk/engine.go.
* **Branch:** main.
* **Notes:** remote bulker map can encounter a race between policy.Prepare output when doing a checkin or ack, and the policy self monitor (that checks for output health). 

https://buildkite.com/elastic/fleet-server/builds/5802#019388a2-40b1-42bf-8b6e-f1c2042fd80f

### Stack Trace

```
=== RUN   Test_Agent_Remote_ES_Output
--
  | remote_es_output_integration_test.go:147: Test fleet server port=37795
  | remote_es_output_integration_test.go:150: Create policy with remote ES output
  | remote_es_output_integration_test.go:181: Create API key and enrollment key for new policy
  | remote_es_output_integration_test.go:214: Enroll agent
  | fleet_integration_test.go:727: map[action:created item:map[access_api_key:X0lPemlKTUJqdThBeUFmVF9iSlU6TW5vQW8tN1VRQ0NjcGxVN3FJVXd6dw== access_api_key_id:_IOziJMBju8AyAfT_bJU actions:<nil> active:true enrolled_at:2024-12-02T18:48:37Z id:84c90f2e-67c0-4179-acfc-32cb048abdbf local_metadata:map[] policy_id:337d887c-628a-42cd-b99b-0c0e14384e68 status:online tags:[] type:PERMANENT user_provided_metadata:<nil>]]
  | remote_es_output_integration_test.go:40: Fake a checkin for agent 84c90f2e-67c0-4179-acfc-32cb048abdbf
  | ==================
  | WARNING: DATA RACE
  | Write at 0x00c0003f4360 by goroutine 3151:
  | runtime.mapassign_faststr()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/runtime/map_faststr.go:223 +0x0
  | github.com/elastic/fleet-server/v7/internal/pkg/bulk.(*Bulker).updateBulkerMap()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/bulk/engine.go:148 +0xef
  | github.com/elastic/fleet-server/v7/internal/pkg/bulk.(*Bulker).CreateAndGetBulker()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/bulk/engine.go:177 +0x64c
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.(*Output).Prepare()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/policy/policy_output.go:64 +0xaa6
  | github.com/elastic/fleet-server/v7/internal/pkg/api.processPolicy()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/handleCheckin.go:833 +0x1027
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*CheckinT).ProcessRequest()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/handleCheckin.go:364 +0x2024
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*CheckinT).handleCheckin()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/handleCheckin.go:145 +0x8de
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*apiServer).AgentCheckin()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/api.go:77 +0x4c4
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*ServerInterfaceWrapper).AgentCheckin.func1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/openapi.gen.go:2179 +0x10a
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*apiVersion).middleware-fm.(*apiVersion).middleware.func1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/apiVersion.go:80 +0x418
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*ServerInterfaceWrapper).AgentCheckin()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/openapi.gen.go:2186 +0x1341
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*ServerInterfaceWrapper).AgentCheckin-fm()
  | <autogenerated>:1 +0x51
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/go-chi/chi/v5.(*ChainHandler).ServeHTTP()
  | /opt/buildkite-agent/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.12/chain.go:31 +0x63
  | github.com/go-chi/chi/v5.(*Mux).routeHTTP()
  | /opt/buildkite-agent/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.12/mux.go:459 +0x594
  | github.com/go-chi/chi/v5.(*Mux).routeHTTP-fm()
  | <autogenerated>:1 +0x51
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/elastic/fleet-server/v7/internal/pkg/limit.(*Limiter).Wrap.func1.1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/limit/limiter.go:94 +0x373
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*limiter).middleware-fm.(*limiter).middleware.func1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/router.go:124 +0x2d5
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/go-chi/chi/v5/middleware.Recoverer.func1()
  | /opt/buildkite-agent/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.12/middleware/recoverer.go:45 +0xf0
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/elastic/fleet-server/v7/internal/pkg/logger.Middleware.func1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/logger/http.go:254 +0xfe7
  | net/http.HandlerFunc.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2220 +0x47
  | github.com/go-chi/chi/v5.(*Mux).ServeHTTP()
  | /opt/buildkite-agent/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.12/mux.go:90 +0x69e
  | net/http.serverHandler.ServeHTTP()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:3210 +0x2a1
  | net/http.(*conn).serve()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:2092 +0x12a4
  | net/http.(*Server).Serve.gowrap3()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:3360 +0x4f
  |  
  | Previous read at 0x00c0003f4360 by goroutine 3088:
  | runtime.mapiterinit()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/runtime/map.go:877 +0x0
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.reportOutputHealth()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/policy/self.go:265 +0xa9
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.(*selfMonitorT).updateState()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/policy/self.go:216 +0x5bc
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.(*selfMonitorT).processPolicies()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/policy/self.go:183 +0x694
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.(*selfMonitorT).Run()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/policy/self.go:122 +0x912
  | github.com/elastic/fleet-server/v7/internal/pkg/policy.SelfMonitor.Run-fm()
  | <autogenerated>:1 +0x61
  | github.com/elastic/fleet-server/v7/internal/pkg/server.loggedRunFunc.func1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/server/fleet.go:303 +0xb8
  | golang.org/x/sync/errgroup.(*Group).Go.func1()
  | /opt/buildkite-agent/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x91
  |  
  | Goroutine 3151 (running) created at:
  | net/http.(*Server).Serve()
  | /opt/buildkite-agent/.gvm/versions/go1.23.3.linux.amd64/src/net/http/server.go:3360 +0x8ec
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*server).Run.func3()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/server.go:122 +0x110
  | github.com/elastic/fleet-server/v7/internal/pkg/api.(*server).Run.gowrap1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/api/server.go:125 +0x82
  |  
  | Goroutine 3088 (running) created at:
  | golang.org/x/sync/errgroup.(*Group).Go()
  | /opt/buildkite-agent/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x124
  | github.com/elastic/fleet-server/v7/internal/pkg/server.(*Fleet).runSubsystems()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/server/fleet.go:504 +0x1898
  | github.com/elastic/fleet-server/v7/internal/pkg/server.(*Fleet).runServer()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/server/fleet.go:429 +0x6ae
  | github.com/elastic/fleet-server/v7/internal/pkg/server.(*Fleet).Run.func4()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/server/fleet.go:199 +0x48
  | github.com/elastic/fleet-server/v7/internal/pkg/server.(*Fleet).Run.func2.1()
  | /opt/buildkite-agent/builds/bk-agent-prod-gcp-1733164329918929620/elastic/fleet-server/internal/pkg/server/fleet.go:128 +0x64
  | golang.org/x/sync/errgroup.(*Group).Go.func1()
  | /opt/buildkite-agent/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x91
  | ==================
  | remote_es_output_integration_test.go:50: Checkin successful, verify body
  | remote_es_output_integration_test.go:110: Fake an ack for action policy:337d887c-628a-42cd-b99b-0c0e14384e68:1 for agent 84c90f2e-67c0-4179-acfc-32cb048abdbf
  | remote_es_output_integration_test.go:129: Ack successful, verify body
  | remote_es_output_integration_test.go:234: Update policy to remove remote ES output
  | remote_es_output_integration_test.go:256: Checkin so that agent gets new policy revision
  | remote_es_output_integration_test.go:40: Fake a checkin for agent 84c90f2e-67c0-4179-acfc-32cb048abdbf
  | remote_es_output_integration_test.go:50: Checkin successful, verify body
  | remote_es_output_integration_test.go:259: Ack so that fleet triggers remote api key invalidate
  | remote_es_output_integration_test.go:110: Fake an ack for action policy:337d887c-628a-42cd-b99b-0c0e14384e68:2 for agent 84c90f2e-67c0-4179-acfc-32cb048abdbf
  | remote_es_output_integration_test.go:129: Ack successful, verify body
  | testing.go:1399: race detected during execution of test


```

### PR DESCRIPTION
**Pull request by michel-laterman**
## What is the problem this PR solves?

Remove a race condition/bug that may occur when remote ES outputs are used.

## How does this PR solve the problem?

Use the `remoteOutputMutex` whenever accessing the `bulkerMap`.
Change `GetBulkerMap` to return a copy of the map so that remote output health will not conflict with adding/removing a bulker from the map.

## Design Checklist

- ~~I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.~~
- ~~I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.~~
- ~~I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.~~

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- ~~I have made corresponding changes to the documentation~~
- ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/fleet-server#changelog)

## Related issues

- Closes #4170

### CODE REVIEW

File: internal/pkg/bulk/engine.go [hunk 1]
```diff
@@ -115,36 +115,46 @@ func NewBulker(es esapi.Transport, tracer *apm.Tracer, opts ...BulkOpt) *Bulker
...
+	o, ok := b.bulkerMap.Load(outputName)
+	if !ok {
+		return nil
+	}
+	return o.(Bulk)
 }
 
+// GetBulkerMap returns a copy of the BulkerMap
 func (b *Bulker) GetBulkerMap() map[string]Bulk {
-	return b.bulkerMap
```
Comment (michel-laterman) [inline internal/pkg/bulk/engine.go:135]: From what I can see, we had a flaky test because the policy self-monitor (`internal/pkg/policy/self.go`) was getting the bulker map in order to check remote output health; but we did not prevent our policy output preparation from creating a new bulker concurrently (`internal/pkg/policy/policy_output.go`).
I've changes our maps to `sync.Map` as we weren't properly using the mutex we had, and changed this func to return a copy instead

File: internal/pkg/bulk/engine.go [hunk 2]
```diff
@@ -115,36 +115,46 @@ func NewBulker(es esapi.Transport, tracer *apm.Tracer, opts ...BulkOpt) *Bulker
...
 	}
 }
 
 func (b *Bulker) GetBulker(outputName string) Bulk {
-	return b.bulkerMap[outputName]
+	o, ok := b.bulkerMap.Load(outputName)
+	if !ok {
+		return nil
+	}
+	return o.(Bulk)
```
Comment (kaanyalti) [inline internal/pkg/bulk/engine.go:132]: Can this panic?

File: internal/pkg/bulk/engine.go [hunk 3]
```diff
@@ -153,12 +163,12 @@ func (b *Bulker) updateBulkerMap(outputName string, newBulker *Bulker) {
 // if changed, stop the existing bulker and create a new one
 func (b *Bulker) CreateAndGetBulker(ctx context.Context, zlog zerolog.Logger, outputName string, outputMap map[string]map[string]interface{}) (Bulk, bool, error) {
 	hasConfigChanged := b.hasChangedAndUpdateRemoteOutputConfig(zlog, outputName, outputMap[outputName])
-	bulker := b.bulkerMap[outputName]
-	if bulker != nil && !hasConfigChanged {
-		return bulker, false, nil
+	bulker, ok := b.bulkerMap.Load(outputName)
+	if ok && !hasConfigChanged {
+		return bulker.(Bulk), false, nil
```
Comment (kaanyalti) [inline internal/pkg/bulk/engine.go:168]: I'm assuming it is guaranteed that the sync.Map will only hold bulker type, is that the case?

Comment (michel-laterman) [reply internal/pkg/bulk/engine.go:168]: yes

File: internal/pkg/bulk/engine.go (hunk 2 above)
Comment (michel-laterman) [reply internal/pkg/bulk/engine.go:132]: it shouldn't, we are only adding Bulkers

**Comment by michel-laterman** [comment]
I added some benchmarks, and ran them against main for comparison.
Note that while main has a `RWMutex`, it's only used in the `updateBulkerMap` func, but not in `GetBulker`, `CreateAndGetBulker`.  
This PR also uses a sync.Map for `remoteOutputConfigMap`, and main does not.

Benchmarks were ran with `go test -bench=Benchmark_CreateAndGetBulker -count 10 .`
```
goos: darwin
goarch: arm64
pkg: github.com/elastic/fleet-server/v7/internal/pkg/bulk
cpu: Apple M3 Pro
                                              │  main.txt   │              sync.txt               │
                                              │   sec/op    │    sec/op     vs base               │
_CreateAndGetBulker/new_remote_bulker-12        44.62µ ± 7%   45.56µ ± 10%       ~ (p=0.529 n=10)
_CreateAndGetBulker/existing_remote_bulker-12   658.5n ± 3%   710.3n ±  6%  +7.86% (p=0.000 n=10)
_CreateAndGetBulker/changed_remote_bulker-12    46.27µ ± 9%   47.56µ ±  5%       ~ (p=0.529 n=10)
geomean                                         11.08µ        11.55µ        +4.22%

                                              │   main.txt   │              sync.txt               │
                                              │     B/op     │     B/op      vs base               │
_CreateAndGetBulker/new_remote_bulker-12        39.30Ki ± 0%   38.79Ki ± 0%  -1.30% (p=0.000 n=10)
_CreateAndGetBulker/existing_remote_bulker-12     591.0 ± 0%     623.0 ± 0%  +5.41% (p=0.000 n=10)
_CreateAndGetBulker/changed_remote_bulker-12    38.78Ki ± 1%   38.78Ki ± 1%       ~ (p=0.896 n=10)
geomean                                         9.582Ki        9.709Ki       +1.32%

                                              │  main.txt   │              sync.txt               │
                                              │  allocs/op  │  allocs/op   vs base                │
_CreateAndGetBulker/new_remote_bulker-12        1.033k ± 0%   1.039k ± 0%   +0.58% (p=0.000 n=10)
_CreateAndGetBulker/existing_remote_bulker-12    11.00 ± 0%    13.00 ± 0%  +18.18% (p=0.000 n=10)
_CreateAndGetBulker/changed_remote_bulker-12    1.038k ± 0%   1.042k ± 0%   +0.39% (p=0.000 n=10)
geomean                                          227.6         241.4        +6.07%
```

File: internal/pkg/bulk/bulk_remote_output_test.go [hunk 4]
```diff
@@ -148,3 +153,57 @@ func Test_CreateAndGetBulkerChanged(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, true, cancelFnCalled)
 }
+
+func Benchmark_CreateAndGetBulker(b *testing.B) {
+	b.Skip("Crashes on remote runner")
```
Comment (michel-laterman) [inline internal/pkg/bulk/bulk_remote_output_test.go:158]: I'm not sure why, but this causes issues when `make benchmark` is ran, but these tests can be ran individually without this

**Review by kaanyalti — APPROVED** [review]

**Comment by michel-laterman** [comment]
buildkite test this

**Comment by cmacknz** [comment]
My take on this is that this is a real bug, the `GetBulkerMap()` method introduced to support remote outputs isn't concurrency safe and really can't be unless you do something like this. 

I don't like the use of `sync.Map` to solve this, primarily because it removes type safety in a codebase that I already find hard to follow with many levels of interface abstraction. I also don't see us satisfying the two criteria where sync.Map recommends its use: https://pkg.go.dev/sync#Map. I also don't like that it will encourage us to just randomly access the bulker everywhere but TBH that is kind of how it is designed to be used right now.

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate [Mutex](https://pkg.go.dev/sync#Mutex) or [RWMutex](https://pkg.go.dev/sync#RWMutex).

We don't satisfy 1 because outputs can be changed, updated, and deleted. We don't satisfy 2 because the entire point of `GetBulkMap` is to look at every key and not some non-overlapping set of keys.

It looks like we could solve this but just introducing a RWMutex for all the uses of bulkMap. There are several that are not mutex protected. https://github.com/elastic/fleet-server/blob/6b29ab474de9bbcbef890da5aeeb07737ffac0d9/internal/pkg/bulk/engine.go#L95


The `GetBulkerMap` method only has one use and it is to get the output name and client for each existing output. Rather than return a reference to the map, you can just hold a mutex to return a copy of those two things. Assuming the client is safe for concurrent access. `GetBulkerMap` shouldn't exist in its current form.

You could potentially rewrite the way the remote ES output healthcheck works to not need concurrency at all. The self monitor could listen on a channel for state updates from each attempt to interact with the remote ES update that could fail or something like that.

**Comment by cmacknz** [comment]
I'm not really worried about the performance implications of this after read the code involved. I think that was just a way to see if use of sync.Map was justified. I don't think it is for non-performance reasons (interfaces, ugh).

**Comment by michel-laterman** [comment]
I'll reimplement the mutex and create an issue to discuss remote output health

**Comment by michel-laterman** [comment]
Issue: https://github.com/elastic/fleet-server/issues/4185
I think we should remove remote health reporting from the policy self-monitor as it does nothing to the (fleet-server) status, but we can discuss it in the issue.
